api/blockindexer/indexer.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package txindexer
+
+import (
+	"sync"
+
+	"github.com/ava-labs/avalanchego/ids"
+
+	"github.com/ava-labs/hypersdk/chain"
+	"github.com/ava-labs/hypersdk/event"
+	"github.com/ava-labs/hypersdk/internal/cache"
+	"github.com/ava-labs/hypersdk/vm"
+)
+
+const (
+	Name      = "blockindexer"
+	Namespace = "blockindexer"
+)
+
+var (
+	_ event.SubscriptionFactory[*chain.StatefulBlock] = (*subscriptionFactory)(nil)
+	_ event.Subscription[*chain.StatefulBlock]        = (*indexer)(nil)
+)
+
+type Config struct {
+	Enabled             bool `json:"enabled"`
+	AcceptedBlockWindow int  `json:"acceptedBlockWindow"`
```

Comment:
1. Why is the `blockindexer` separate from the `txindexer`? It seems reasonable to just have a single `indexer` component that can enable either/or indexing for txs/blocks.

2. Why is an in-memory indexer useful? If you have a downstream component reading from the indexer, any downtime in the downstream component can risk reliability loss.
    * If your downstream component is down for some time and > `acceptedBlockWindow` blocks are produced, you have no way of recovering the blocks lost between what downstream last read and the oldest block in the indexer.
    * Would it be more reasonable for the indexer be able to configure a period of time for which it keeps blocks on disk? If we did that we could just say that the downstream component is responsible for not falling behind the server's configured data retention.

Reply:
1. I thought it was cleaner to separate them, but can re-combine into one if you prefer
2. This API explicitly chooses not to provide reliability for blocks. This was added to provide the minimum API for https://github.com/ava-labs/hypersdk/issues/1513.

We talked about this previously, but the best argument imo for implementing a persistent rolling index is to support running it both in and out of the node. It seemed too strange to me that the VM cleans up blocks as it moves forward and then a separate API keeps its own persistent index.